validate/src/request.rs

```diff
@@ -315,14 +347,29 @@ impl Display for ValidationError {
 
                 Display::fmt(&TEMPLATE_NAME_LENGTH_MAX, f)
             }
-            ValidationErrorType::Username { len } => {
-                f.write_str("provided username length is ")?;
-                Display::fmt(len, f)?;
-                f.write_str(", but it must be at least ")?;
-                Display::fmt(&USERNAME_LIMIT_MIN, f)?;
-                f.write_str(" and at most ")?;
-
-                Display::fmt(&USERNAME_LIMIT_MAX, f)
+            ValidationErrorType::Username { len, substring }
+            | ValidationErrorType::WebhookUsername { len, substring } => {
+                f.write_str("provided username ")?;
+
+                if let Some(len) = len {
+                    f.write_str("length is ")?;
+                    Display::fmt(len, f)?;
+                    f.write_str(", but it must be at least ")?;
+                    Display::fmt(&USERNAME_LIMIT_MIN, f)?;
+                    f.write_str(" and at most ")?;
+                    Display::fmt(&USERNAME_LIMIT_MAX, f)?;
+                }
+
+                if let Some(substring) = substring {
+                    if len.is_some() {
+                        f.write_str(", and")?;
+                    }
+
+                    f.write_str(" cannot contain ")?;
+                    Display::fmt(substring, f)?;
+                }
+
+                Ok(())
             }
```

Comment:
This logic seems a bit fuzzy to me. I went like "oh, but if this isn't present then it'll read weird.. ah, no.. but if that's not present.. right." How about a test to get it written that this is for sure correct? :)